// @cloudbase/ai/dist/cjs/models/image-model.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultImageModel = void 0;
var errors_1 = require("../errors");
var DefaultImageModel = (function () {
    function DefaultImageModel(req, baseUrl, provider) {
        this.req = req;
        this.baseUrl = baseUrl;
        this.provider = provider;
        this.defaultGenerateImageSubUrl = 'images/ar/generations';
        this.generateImageSubUrlConfig = {
            'hunyuan-image': [[/^hunyuan-image(-v[12](\..*)?)?$/, 'images/generations']],
        };
        this.generateImage = this.generateImageImpl.bind(this);
    }
    DefaultImageModel.prototype.generateImageImpl = function (input) {
        return __awaiter(this, void 0, void 0, function () {
            var typedInput, imageUrls, images, subUrl, newSubUrl, res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (typeof input === 'object' && input && 'model' in input && input.model === 'HY-Image-v3.0-I2I-ToB-v1.0.1'
                            && (this.provider === '' || this.provider === 'hunyuan-image')) {
                            typedInput = input;
                            imageUrls = typedInput.image_urls;
                            images = typedInput.images;
                            if (imageUrls && imageUrls.length > 1) {
                                throw new errors_1.InvalidParamsError('当模型为 HY-Image-v3.0-I2I-ToB-v1.0.1 时，image_urls 最多只能有 1 个元素', 'image_urls');
                            }
                            if (images && images.length > 1) {
                                throw new errors_1.InvalidParamsError('当模型为 HY-Image-v3.0-I2I-ToB-v1.0.1 时，images 最多只能有 1 个元素', 'images');
                            }
                        }
                        subUrl = this.defaultGenerateImageSubUrl;
                        if (typeof input === 'object' && input && 'model' in input && typeof input.model === 'string') {
                            newSubUrl = this.getGenerateImageSubUrl(this.provider, input.model);
                            if (newSubUrl)
                                subUrl = newSubUrl;
                        }
                        return [4, this.req({
                                url: this.getGenerateUrl(subUrl),
                                data: input,
                                stream: false,
                            })];
                    case 1:
                        res = (_a.sent());
                        return [2, res];
                }
            });
        });
    };
    DefaultImageModel.prototype.getGenerateUrl = function (subUrl) {
        return "".concat(this.baseUrl, "/").concat(this.provider, "/").concat(subUrl);
    };
    DefaultImageModel.prototype.getGenerateImageSubUrl = function (provider, model) {
        var rules = this.generateImageSubUrlConfig[provider];
        if (!rules)
            return undefined;
        for (var _i = 0, rules_1 = rules; _i < rules_1.length; _i++) {
            var _a = rules_1[_i], pattern = _a[0], subUrl = _a[1];
            if (pattern.test(model))
                return subUrl;
        }
        return undefined;
    };
    return DefaultImageModel;
}());
exports.DefaultImageModel = DefaultImageModel;
